not compatible with c++20

std::allocator::rebind was deprecated in c++17, and removed in c++20. It should be replaced with std::allocator_traits::rebind_alloc. The use of allocator rebind means when compiling with GCC10 and -std=c++20, compilation fails with something like

```
In file included from /usr/local/include/s2/s2shape_index.h:44,
                 from /usr/local/include/s2/mutable_s2shape_index.h:37,
                 from /usr/local/include/s2/s2polygon.h:29,
                 from ...:
/usr/local/include/s2/util/gtl/compact_array.h: In instantiation of 'class gtl::compact_array_base<S2ClippedShape, std::allocator<S2ClippedShape> >':
/usr/local/include/s2/util/gtl/compact_array.h:520:7:   required from 'class gtl::compact_array<S2ClippedShape>'
/usr/local/include/s2/s2shape_index.h:157:21:   required from here
/usr/local/include/s2/util/gtl/compact_array.h:148:49: error: no class template named 'rebind' in 'class std::allocator<S2ClippedShape>'
  148 |   typedef typename A::template rebind<T>::other value_allocator_type;
      |                                                 ^~~~~~~~~~~~~~~~~~~~
In file included from /usr/local/include/s2/util/gtl/dense_hash_set.h:128,
                 from /usr/local/include/s2/sequence_lexicon.h:26,
                 from /usr/local/include/s2/id_set_lexicon.h:26,
                 from /usr/local/include/s2/s2builder.h:30,
                 from /usr/local/include/s2/s2boolean_operation.h:24,
                 from /usr/local/include/s2/s2polygon.h:31,
```

--- a/src/s2/util/gtl/compact_array.h
+++ b/src/s2/util/gtl/compact_array.h
@@ -145,7 +145,8 @@ class compact_array_base {
     return const_cast<compact_array_base<T, A>*>(this)->Array();
   }
 
-  typedef typename A::template rebind<T>::other value_allocator_type;
+  using value_allocator_type =
+      typename std::allocator_traits<A>::template rebind_alloc<T>;
 
  public:
   typedef T                                     value_type;

--- a/src/s2/util/gtl/densehashtable.h
+++ b/src/s2/util/gtl/densehashtable.h
@@ -182,7 +182,8 @@ struct dense_hashtable_const_iterator;
 template <class V, class K, class HF, class ExK, class SetK, class EqK, class A>
 struct dense_hashtable_iterator {
  private:
-  typedef typename A::template rebind<V>::other value_alloc_type;
+  using value_alloc_type =
+      typename std::allocator_traits<A>::template rebind_alloc<V>;
 
  public:
   typedef dense_hashtable_iterator<V, K, HF, ExK, SetK, EqK, A>
@@ -245,7 +246,8 @@ struct dense_hashtable_iterator {
 template <class V, class K, class HF, class ExK, class SetK, class EqK, class A>
 struct dense_hashtable_const_iterator {
  private:
-  typedef typename A::template rebind<V>::other value_alloc_type;
+  using value_alloc_type =
+      typename std::allocator_traits<A>::template rebind_alloc<V>;
 
  public:
   typedef dense_hashtable_iterator<V, K, HF, ExK, SetK, EqK, A>
@@ -311,7 +313,8 @@ template <class Value, class Key, class HashFcn,
           class ExtractKey, class SetKey, class EqualKey, class Alloc>
 class dense_hashtable {
  private:
-  typedef typename Alloc::template rebind<Value>::other value_alloc_type;
+  using value_alloc_type =
+      typename std::allocator_traits<Alloc>::template rebind_alloc<Value>;
 
 
  public:

